### ISSUE TITLE
json vs xml

### ISSUE THREAD

**Issue by skadefro** [object/Simple.OData.Client#81]
hey
Is there anyway to change the accept header for SOC ?
currently it sends application/json, application/xml, application/text
but with the newest version of web.api odata (v4) application/xml is nolonger valid, only "application/atom+xml".
And for some reason, everytime SOC sees json it just plain fails with "No property or association found for XXXX" so would be nice if i could force it back to using xml.

**Comment by object** [comment]
Did you try version 4.2.1? It reorders the accept headers prioritizing application/json before others. Otherwise it's possible to change the headers in BeforeRequest interceptor.

**Comment by skadefro** [comment]
hey

sorry for the late reply, suddenly had to focus on something else for a while.
Downloaded and compiled 4.2.1.0 but didnt change much. but using 

                        var options = new Simple.OData.Client.ODataClientSettings()
                        {
                            UrlBase = ccp.db.baseurl,
                            BeforeRequest = (r) => {
                                r.Headers.Accept.Clear();
                                r.Headers.Accept.Add(new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/atom+xml"));
                            }
                        };


worked perfectly.

If i want to use json instead, is there anyway to get more useful error information about why it keeps failing on serializing with "No property or association found for XXXX" where xxx is any random property ?

**Comment by object** [comment]
Hmm, you should be able to use json instead. I would like to get more details about the error if possible.

**Comment by skadefro** [comment]
After swapping around versions, I'm now stuck. Doing updates it's always sending in json, and getting the result back in json too, resulting in the dreaded "No property or association found", so yes any help would be greatly appreciated. 
What do you need ? or should could we do a quick teamviewer session so you can see it ?

**Comment by object** [comment]
If you share with me a sample project (or give me an access to your OData feed) as well as client code you are using, I can check this out. It all seems strange to me because S.O.D. v4 is widely used with JSON payload.

**Comment by skadefro** [comment]
hey.

I triede several times to "split" it up, but gave up, so here is just "everything".
if you run the project wftestlab, you'll see the error right away.

Please comment/email when you downloaded it, so i can remove it again.
http://admin.wingu.dk/CCP_3.zip

**Comment by object** [comment]
Great, thanks! I will have a look.

**Comment by skadefro** [comment]
Sorry, sorry. When zipping it, I did a clean to remove binaries. and forgot to rebuild, so OData wasn't working. I've rebuilt it now...

**Comment by object** [comment]
The file is not available. Perhaps you already removed it. Please upload it again.

**Comment by skadefro** [comment]
yes, i have uploaded it again, same url.

**Comment by object** [comment]
Got it, thanks!

**Comment by object** [comment]
I tested your project, and apparently the issue is not related to json vs. xml, it's something else. "No property or association found for instanceid" is raised from this code:

var t = e.save<ccp.entities.wfevent>();

"wfevent" refers to instanceid that is declared in "wfevent" OData EntityType. But I can't see any EntitySet that is derived from that type. The library tried to find EntitySet with the name "wfevent" or "wfevents" and it finds "wfevents" of course. But that one doesn't have "instanceid" field, so you are getting this error.

**Comment by skadefro** [comment]
... Sigh ... why you always find embarrassing bugs in my code 
You were right; the entity set was exposed with the wrong class (and name)

            //var wfevents = builder.EntitySet<ccp.entities.wf>("wfevents");
            var wfevents = builder.EntitySet<ccp.entities.wfevent>("wfevent");

Thank you :)

**Comment by object** [comment]
No problem! Four eyes is always better than two. BTW when zipping projects to share, it's OK not to include "packages" folder. Packages are restored automatically but they take a lot of space.

**Comment by rasikas** [comment]
I have a question related to using json instead of xml in oData calls, I thought to ask the question in this comment instead of opening a new issue. I want to communicate using json in order to minimize the data traffic. So I used following settings when initializing odata client

~~~
            var settings = new ODataClientSettings("http://services.odata.org/northwind/northwind.svc/");
            settings.BeforeRequest = (h) =>
            {   //Content-Type: application/json; odata.metadata=minimal                
                h.Headers.Accept.Clear();
                h.Headers.Accept.Add(new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));
            };
            var client = new ODataClient(settings);
            var categories = await client.For<Category>().FindEntriesAsync();
~~~

First question, is this the proper way to instruct oDataClient to use json instead of xml.
I found that client throws an exception when I use this because oDataClient go and fetcx $metadata which needs to be in xml format.

Exception: {"Unsupported Media Type"}
message=A supported MIME type could not be found that matches the acceptable MIME types for the request. The supported type(s) 'application/xml' do not match any of the acceptable MIME types 'application/json'.

So how do we setup the client to use 'application/xml' in its metadata call (which it does automatically) and to use json for data calls? thank you
